### PR TITLE
add access token authentication

### LINKED ISSUE #65: All Access Tokens for Authentication
### What problem are you facing?
Currently in a GCP environment that restricts the creation of service account keys.

### How could Official GCP Provider help solve your problem?
Would like to have the ability to use temporary access tokens that can be generated via a CronJob for provider authentication.

### PR DESCRIPTION
**Pull request by bradkwadsworth-mw**
Signed-off-by: Brad Wadsworth <brad.wadsworth@mavenwave.com>

### Description of your changes
Allow use of temporary service account access tokens to be used for authentication.

Fixes #65 

I have:

- [X] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
Updated [Configuration.md](https://github.com/upbound/provider-gcp/blob/main/docs/Configuration.md) with instructions on how to test with access tokens.

### CODE REVIEW

**Review by ulucinar — COMMENTED** [review]
Hi @bradkwadsworth-mw,
Thank you for working on introducing token-based authentication support to provider-gcp. 

My understanding is that the PR proposes to switch to token-based auth. by inspecting the referenced credential data (trying to parse the supplied credentials as JSON) and if this succeeds, use token authentication in the underlying Terraform (native) provider by configuring the `access_token` argument in the provider block. And if it fails, it configures the `credentials` argument of the provider block for the native provider.

I would prefer a more explicit API where we have a separate credential source (enum) constant for the token-based authentication like we do for provider-aws [here](https://github.com/upbound/provider-aws/blob/411cb2b7d60ef6dec1118208c32371f3038cf3d3/internal/clients/provider_config.go#L39). Then we could configure the `access_token` argument for the native provider configuration block in [this](https://github.com/upbound/provider-gcp/blob/e0252bbec3e86726f82df44ffb91df9c6fcb7076/internal/clients/gcp.go#L77) switch when the credential source is `AccessToken`. 

This would expose an explicit API for token-based authentication with, in my opinion, better UX. Because we are surfacing the API to the credential source (enums) level, it will be easier for users to follow what options are supported and to enable them to specify their intentions (regarding the auth. mechanism to use) in a more explicit manner. For instance, if the user of the `ProviderConfig` API would like to use token-based auth. but has a syntax error in the JSON credentials she supplies, then we would implicitly be switching to service principal credentials-based authentication, although the intent was different. What do you think?

**Comment by bradkwadsworth-mw** [comment]
> Hi @bradkwadsworth-mw, Thank you for working on introducing token-based authentication support to provider-gcp.
> 
> My understanding is that the PR proposes to switch to token-based auth. by inspecting the referenced credential data (trying to parse the supplied credentials as JSON) and if this succeeds, use token authentication in the underlying Terraform (native) provider by configuring the `access_token` argument in the provider block. And if it fails, it configures the `credentials` argument of the provider block for the native provider.
> 
> I would prefer a more explicit API where we have a separate credential source (enum) constant for the token-based authentication like we do for provider-aws [here](https://github.com/upbound/provider-aws/blob/411cb2b7d60ef6dec1118208c32371f3038cf3d3/internal/clients/provider_config.go#L39). Then we could configure the `access_token` argument for the native provider configuration block in [this](https://github.com/upbound/provider-gcp/blob/e0252bbec3e86726f82df44ffb91df9c6fcb7076/internal/clients/gcp.go#L77) switch when the credential source is `AccessToken`.
> 
> This would expose an explicit API for token-based authentication with, in my opinion, better UX. Because we are surfacing the API to the credential source (enums) level, it will be easier for users to follow what options are supported and to enable them to specify their intentions (regarding the auth. mechanism to use) in a more explicit manner. For instance, if the user of the `ProviderConfig` API would like to use token-based auth. but has a syntax error in the JSON credentials she supplies, then we would implicitly be switching to service principal credentials-based authentication, although the intent was different. What do you think?

That sounds good to me as well. I'll work on that.

**Comment by ulucinar** [comment]
/test-examples="examples/compute/address.yaml"

File: internal/clients/gcp.go [hunk 1]
```diff
@@ -77,15 +79,20 @@ func TerraformSetupBuilder(version, providerSource, providerVersion string) terr
 		switch pc.Spec.Credentials.Source { //nolint:exhaustive
 		case xpv1.CredentialsSourceInjectedIdentity:
 			// We don't need to do anything here, as the TF Provider will take care of workloadIdentity etc.
+		case accessToken:
+			data, err := resource.CommonCredentialExtractor(ctx, "Secret", client, pc.Spec.Credentials.CommonCredentialSelectors)
```
Comment (ulucinar) [inline internal/clients/gcp.go:83]: nit:
```suggestion
			data, err := resource.CommonCredentialExtractor(ctx, xpv1.CredentialsSourceSecret, client, pc.Spec.Credentials.CommonCredentialSelectors)
```

File: internal/clients/gcp.go [hunk 2]
```diff
@@ -77,15 +79,20 @@ func TerraformSetupBuilder(version, providerSource, providerVersion string) terr
...
+			if err != nil {
+				return ps, errors.Wrap(err, errExtractCredentials)
```
Comment (ulucinar) [inline internal/clients/gcp.go:85]: nit:
We had better differentiate the message of the error returned here from the `default` case's error.

**Review by ulucinar — COMMENTED** [review]
Hi @bradkwadsworth-mw,
Thank you very much for working on the new authentication scheme!
Left some nitpicks but I think the most important review comment is regarding the import statement. Because the actual authentication is performed by the underlying Terraform provider, which is a separate binary, do we really need the import statement?

File: cmd/provider/main.go [hunk 3]
```diff
@@ -45,6 +45,8 @@ import (
 	"github.com/upbound/provider-gcp/internal/clients"
 	"github.com/upbound/provider-gcp/internal/controller"
 	"github.com/upbound/provider-gcp/internal/features"
+
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
```
[the remark this answers is not in the document]
Comment (bradkwadsworth-mw) [reply cmd/provider/main.go:49]: This import was required in order to get `make run` to work. Without it `make run` will throw this error.
```
UPBOUND_CONTEXT="local" /Users/brad/Workspace/crossplane/upbound-provider-gcp/_output/bin/darwin_amd64/provider --debug
1.672953250440712e+09   DEBUG   provider-gcp    Starting        {"sync-interval": "1h0m0s", "poll-interval": "10m0s", "max-reconcile-rate": 10}
1.67295325044181e+09    ERROR   Failed to get API Group-Resources       {"error": "no Auth Provider found for name \"gcp\""}
sigs.k8s.io/controller-runtime/pkg/cluster.New
        sigs.k8s.io/controller-runtime@v0.12.1/pkg/cluster/cluster.go:160
sigs.k8s.io/controller-runtime/pkg/manager.New
        sigs.k8s.io/controller-runtime@v0.12.1/pkg/manager/manager.go:322
main.main
        github.com/upbound/provider-gcp/cmd/provider/main.go:85
runtime.main
        runtime/proc.go:250
provider: error: Cannot create controller manager: no Auth Provider found for name "gcp"
make: *** [run] Error 1
```

**Comment by ulucinar** [comment]
Triggered to check the Secret provider authentication scheme:
/test-examples="examples/compute/address.yaml"

File: cmd/provider/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ulucinar) [reply cmd/provider/main.go:49]: Hi @bradkwadsworth-mw,
I've given `make run` a try after removing the import as follows:
```shell
diff --git a/cmd/provider/main.go b/cmd/provider/main.go
index cfe416d..f24bdaf 100644
--- a/cmd/provider/main.go
+++ b/cmd/provider/main.go
@@ -45,8 +45,6 @@ import (
        "github.com/upbound/provider-gcp/internal/clients"
        "github.com/upbound/provider-gcp/internal/controller"
        "github.com/upbound/provider-gcp/internal/features"
-
-       _ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
 )

 func main() {
diff --git a/go.mod b/go.mod
index 8f1e27a..8d96f4d 100644
--- a/go.mod
+++ b/go.mod
@@ -16,7 +16,6 @@ require (
 )

 require (
-       cloud.google.com/go v0.97.0 // indirect
        github.com/PuerkitoBio/purell v1.1.1 // indirect
        github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578 // indirect
        github.com/agext/levenshtein v1.2.3 // indirect
```

Looks like I can run the provider with `make run` against a `kind` cluster:
```shell
> make run
...
1.673360088962421e+09   DEBUG   provider-gcp    plan ended      {"workspace": "/var/folders/ws/3fn5p0_d45v6vx_z1_sxjxm80000gn/T/d70627b7-c76f-4633-8d06-4362828fc8ad", "out": "{\"@level\":\"info\",\"@message\":\"Terraform 1.2.1\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2023-01-10T17:14:47.847046+03:00\",\"terraform\":\"1.2.1\",\"type\":\"version\",\"ui\":\"1.0\"}\n{\"@level\":\"info\",\"@message\":\"Plan: 0 to add, 0 to change, 0 to destroy.\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2023-01-10T17:14:48.960953+03:00\",\"changes\":{\"add\":0,\"change\":0,\"remove\":0,\"operation\":\"plan\"},\"type\":\"change_summary\"}\n"}
1.673360088962485e+09   DEBUG   provider-gcp    External resource is up to date {"controller": "managed/compute.gcp.upbound.io/v1beta1, kind=subnetwork", "request": "/example-ip-address-subnet", "uid": "d70627b7-c76f-4633-8d06-4362828fc8ad", "version": "1691", "external-name": "example-ip-address-subnet", "requeue-after": 1673360688.962484}
```

```shell
> k get managed
NAME                                                          READY   SYNCED   EXTERNAL-NAME               AGE
subnetwork.compute.gcp.upbound.io/example-ip-address-subnet   True    True     example-ip-address-subnet   6m19s

NAME                                                             READY   SYNCED   EXTERNAL-NAME                     AGE
address.compute.gcp.upbound.io/example-ip-address-with-network   True    True     example-ip-address-with-network   6m19s

NAME                                                        READY   SYNCED   EXTERNAL-NAME                AGE
network.compute.gcp.upbound.io/example-ip-address-network   True    True     example-ip-address-network   6m19s
```

I'm just trying to understand why we need to import this. Looking at the code changes what we do is to read the access token from the K8s secret and put a new Terraform provider configuration option (i.e., the `access_token`), it's the Terraform provider itself using this access token to authenticate against the GCP API. So, it looks like to me that there should be no need for the import.

What happens when you try to `make run` from the main branch HEAD? What type of cluster are you trying to run the provider against? Could this be related to your kubeconfig? Could you please try this with a kind cluster?

[the remark this answers is not in the document]
Comment (bradkwadsworth-mw) [reply cmd/provider/main.go:49]: This does work with a Kind cluster, but does not work when testing against a GKE cluster. You need to test against a GKE cluster in order to fully test out the workload identity component which is required to fetch the temporary access token. I get the same error when running a `make run` against  HEAD and using a GKE cluster.

[the remark this answers is not in the document]
Comment (ulucinar) [reply cmd/provider/main.go:49]: Thank you @bradkwadsworth-mw for bearing with me, totally makes sense.

**Review by ulucinar — APPROVED** [review]

**Pull request merged**